### PR TITLE
[Feature, DEV-142] add draft ERC-884 token implementation

### PR DESCRIPTION
**Pull request by davesag**
also upgraded solidity coverage

The token contract is pretty raw right now and needs some polish, but this at least is a start.

### CODE REVIEW

File: contracts/CTKNToken.sol [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    modifier isNotStockholder(address addr) {
+        require(holderIndices[addr] == 0);
+        _;
+    }
+
+    modifier isNotCancelled(address addr) {
+        require(cancellations[addr] == ZERO_ADDRESS);
+        _;
+    }
+
```
Comment (tavakyan) [inline contracts/CTKNToken.sol:59]: no need for isCancelled?

File: contracts/CTKNToken.sol [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    /**
+     * As each token is minted it is added to the stockholders array.
+     * @param _to The address that will receive the minted tokens.
+     * @param _amount The amount of tokens to mint.
+     * @return A boolean that indicates if the operation was successful.
+     */
+    function mint(address _to, uint256 _amount)
+        public
+        onlyOwner
```
Comment (tavakyan) [inline contracts/CTKNToken.sol:68]: We can later change this to onlyWallet for multisig if we wished.  Also, was wondering should we use CappedToken? Its mintable: https://github.com/OpenZeppelin/zeppelin-solidity/blob/master/contracts/token/ERC20/CappedToken.sol

File: contracts/CTKNToken.sol [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+     *  unknown to the contract then this does nothing. If the address is successfully removed this
+     *  function must emit `VerifiedAddressRemoved(addr, msg.sender)`.
+     *  It MUST throw if an attempt is made to remove a verifiedAddress that owns Tokens.
+     *  @param addr The verified address to be removed.
+     */
+    function removeVerified(address addr)
+        public
+        onlyOwner
+    {
+        require(balances[addr] == 0);
```
Comment (tavakyan) [inline contracts/CTKNToken.sol:140]: Do we need require(addr != ZERO_ADDRESS); before this line?

File: contracts/CTKNToken.sol [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+     *  @param hash A cryptographic hash of the address holder's verified information.
+     */
+    function addVerified(address addr, bytes32 hash)
+        public
+        onlyOwner
+        isNotCancelled(addr)
+    {
+        require(addr != ZERO_ADDRESS);
+        require(hash != ZERO_BYTES);
+        require(verified[addr] == ZERO_BYTES);
```
Comment (tavakyan) [inline contracts/CTKNToken.sol]: How come not verified[addr] != ZERO_BYTES or perhaps the modifier isVerified?

File: contracts/CTKNToken.sol [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        onlyOwner
+        isStockholder(original)
+        isNotStockholder(replacement)
+        isVerifiedAddress(replacement)
+    {
+        // replace the original address in the stockholders array
+        // and update all the associated mappings
+        verified[original] = ZERO_BYTES;
+        cancellations[original] = replacement;
+        uint256 holderIndex = holderIndices[original] - 1;
```
Comment (tavakyan) [inline contracts/CTKNToken.sol]: maybe it should be called holderShares?

Comment (tavakyan) [reply contracts/CTKNToken.sol]: Not sure if I understand what holderIndices is.  Is it the number of shares for a given a stock holder.

Comment (tavakyan) [reply contracts/CTKNToken.sol]: Nevermind, I get it.

File: contracts/CTKNToken.sol [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+     *  transfer or transferFrom will reduce their balance to 0, then
+     *  we need to remove them from the stockholders array.
+     *  @param addr The address to prune if their balance will be reduced to 0.
+     @  @dev see https://ethereum.stackexchange.com/a/39311
+     */
+    function pruneStockholders(address addr, uint256 value)
+        internal
+    {
+        uint256 balance = balances[addr] - value;
+        if (balance > 0) {
```
Comment (tavakyan) [inline contracts/CTKNToken.sol]: Guessing you don't want it to throw with: require ( balances[addr] != value)?

File: contracts/CTKNToken.sol (hunk 5 above)
Comment (tavakyan) [inline contracts/CTKNToken.sol]: can we use two internal functions in here to make it more readable? 

updateHolderIndex(original, replacement)
updateBalances(original, replacement)
?

Comment (tavakyan) [reply contracts/CTKNToken.sol]: Unless it must be atomic / all in one, just a bit hard to follow.

**Review by tavakyan — CHANGES_REQUESTED** [review]
The only issue I think was cancelandReissue was a bit tricky to read.

File: contracts/CTKNToken.sol (hunk 2 above)
Comment (davesag) [reply contracts/CTKNToken.sol:68]: We don't have the need for a token `cap`

File: contracts/CTKNToken.sol (hunk 3 above)
Comment (davesag) [reply contracts/CTKNToken.sol:140]: no because `balances[0x0]` must be `0`

File: contracts/CTKNToken.sol (hunk 1 above)
Comment (davesag) [reply contracts/CTKNToken.sol:59]: nope

File: contracts/CTKNToken.sol (hunk 5 above)
Comment (davesag) [reply contracts/CTKNToken.sol]: I don't think breaking it down into internal functions would help but I will add more comments to make it clearer what is going on.

**Review by tavakyan — APPROVED** [review]
